tensorflow/core/kernels/segment_reduction_ops_impl_1.cc

```diff
@@ -125,21 +125,26 @@ REGISTER_COMPLEX_CPU_KERNELS_ALL(complex128);
           type, index_type,                                                  \
           functor::SegmentReductionFunctor<                                  \
               type, index_type, initial_value_functor,                       \
-              reduction_kernel_functor, atomic_reduction_kernel_functor> >)
+              reduction_kernel_functor, atomic_reduction_kernel_functor>,    \
+          deterministic_for_float>)
 
 #define REGISTER_GPU_SORTED_KERNELS(type, index_type)                     \
   REGISTER_GPU_KERNEL_SORTEDSEGMENT(                                      \
       "SegmentSum", type, index_type, functor::Zero<type>,                \
-      functor::NonAtomicSumOpGpu<type>, functor::AtomicSumOpGpu<type>);   \
+      functor::NonAtomicSumOpGpu<type>, functor::AtomicSumOpGpu<type>,    \
+      /*deterministic_for_float=*/false);                                 \
   REGISTER_GPU_KERNEL_SORTEDSEGMENT(                                      \
       "SegmentProd", type, index_type, functor::One<type>,                \
-      functor::NonAtomicProdOpGpu<type>, functor::AtomicProdOpGpu<type>); \
+      functor::NonAtomicProdOpGpu<type>, functor::AtomicProdOpGpu<type>,  \
+      /*deterministic_for_float=*/false);                                 \
   REGISTER_GPU_KERNEL_SORTEDSEGMENT(                                      \
       "SegmentMin", type, index_type, functor::Highest<type>,             \
-      functor::NonAtomicMinOpGpu<type>, functor::AtomicMinOpGpu<type>);   \
+      functor::NonAtomicMinOpGpu<type>, functor::AtomicMinOpGpu<type>,    \
+      /*deterministic_for_float=*/true);                                  \
   REGISTER_GPU_KERNEL_SORTEDSEGMENT(                                      \
       "SegmentMax", type, index_type, functor::Lowest<type>,              \
-      functor::NonAtomicMaxOpGpu<type>, functor::AtomicMaxOpGpu<type>);
+      functor::NonAtomicMaxOpGpu<type>, functor::AtomicMaxOpGpu<type>,    \
+      /*deterministic_for_float=*/true);
```

Comment:
Can the `OpKernel`s infer `deterministic_for_float` from the `AtomicFooOpGpu` functors?

Reply:
I wish they could do. Without completely re-writing a lot of this code, I couldn't figure out a way of doing this the same way for the unsorted ops and for the sorted ops because the code is structured differently for the two sets of ops. For the sorted ops, the reduction functor types bypass the GPU-specific code and appear in non-device-specific code. If I moved the exception-testing into that code, I would need to condition it on device. This repetition-inducing asymmetry has been bothering me too. Let me take a look at this again and see if I can clean it up using a different approach.

Reply:
It looks like I misunderstood something about the structure of the code. I think this can actually be cleaned up pretty easily.

Reply:
Okay, I'm back to understanding the actual problem that the submitted code solved. The place where we would ideally throw the exception for the sorted segment reduction ops (in order to use the `is_associative` state from `AtomicFooOpGpu`) is running asynchronous compute but doesn't have access to the appropriate `DoneCallBack` in order to call `OP_REQUIRES_ASYNC`. Trying to solve this another way ...

Reply:
I think this is good enough now. Please let me know if you can see a cleaner way of doing it.